### ISSUE TITLE
Future of SDL's gesture APIs

### ISSUE THREAD

**Issue by slime73** [libsdl-org/SDL#6758]
There was a forum thread made a few months ago about what to do with SDL's touch gesture API. I haven't found an equivalent github issue about it so I figured I'd make one in case it's forgotten (although I suppose there might be internal discussions about it, but here it is anyway).

Here's the forum thread https://discourse.libsdl.org/t/sdl-gesture-api/19417

My own two cents: I feel like there are four main possibilities: keep it as-is, remove the gesture API completely, split it into a separate library, or change it so the main SDL API exposes [native OS touch gestures](https://developer.apple.com/documentation/uikit/uigesturerecognizer?language=objc) instead of a custom system.

While I'm not totally familiar with how consistent native gesture support is across operating systems, if gestures continue to exist as a concept in SDL I'd personally prefer to have native gestures exposed rather than alternatives, because it fits better with what SDL is already doing with other APIs, and (I think) it's harder to do in third party code than it would be inside SDL.

I've never used SDL2's gesture API in my own code - although I've had users of my stuff that uses SDL request native OS gestures before - so I only have a limited perspective on this.

I'm not entirely sure from the wording of the issue but #6137 might be related.

**Comment by icculus** [comment]
Honestly, I assumed we were removing it.

The gesture stuff was written by a Google Summer of Code student a long time ago and has been largely unmaintained since. I'm pretty sure I never used it beyond playing with a demo of it for that Summer of Code project, so I have a very low familiarity with it.

I imagine _someone_ is using it in the wide world, but I've never come across it.

My instinct is it should get ripped out of SDL3, and moved wholesale into sdl2-compat; since it isn't a large piece of code and there isn't any platform-specific code in it, it should drop right in, I think.

If someone wants to move to SDL3 _and_ wants the gesture API, I'm not against it becoming a public-domain header-only library they can just include in their project, for which we make no formal releases, and beyond accepting pull requests we do no development on.

**Comment by 1bsyl** [comment]
I am not using it neither (but I remember I did try it). But it sounds a very nice feature to me. 
All the more important, as linux/window/ios/android can have touch screen ...
It makes gesture be available on platform without the need of any specif back-end.


It could be better integrated.
Currently, it always process touch event ( SDL_GestureProcessEvent() in SDL_PushEvent)

Instead, use should somehow "enable" it, and then it should register itself as watch_event callback.

And for instance, if you have no touch screen, you should be able to configure your "mouse_event" be translated as "touch event" (already possible). And then trigger a gesture based on it.

**Comment by icculus** [comment]
My preference is still to move it to an external header-only library, but to be fair, it's not a giant chunk of code if we decide to keep it.

**Comment by slouken** [comment]
I have never seen anyone use this functionality. I think it's one of those things we can remove early in SDL3 and re-add if it turns out that this is a popular feature.

**Comment by icculus** [comment]
I'm taking this one, so I can deal with sliding it into sdl2-compat at the same time.

**Comment by 1bsyl** [comment]
But it's a pity not to have basic event gestures like 
SDL_EVENT_GESTURE_PINCH_IN
SDL_EVENT_GESTURE_PINCH_OUT
SDL_EVENT_GESTURE_SCROLL

(btw, this seems to use 
https://github.com/urho3d/Urho3D/issues/304 
https://github.com/urho3d/Urho3D/blob/master/Source/Urho3D/Input/Input.cpp#L2128
otherwise to many unrelevant occurrence on github to find something..)

**Comment by slime73** [comment]
> But it's a pity not to have basic event gestures like `SDL_EVENT_GESTURE_PINCH_IN`, `SDL_EVENT_GESTURE_PINCH_OUT`, `SDL_EVENT_GESTURE_SCROLL`

I think having an API that exposes native OS touch gestures could be useful, but only after the existing gesture API / implementation is removed otherwise it'd be pretty confusing to people IMO.

**Comment by 1bsyl** [comment]
but that's more work ..
for android, ios, wayland, x11, vita,  rpy ? not sure which are all others...
then what we should use also on x11/wayland ? use libinput ? or reimplement SDL_gesture.c ...
on vita ? is there even some lib for it ?

non-native api has the advantage to be more portable and consistent across platform

but the advantage also of native api I see, is that you implement some (inventing) "gesture_show_desktop".  and it's a 4 fingers on android, and 3 finger on ios, then this is more native behavior ...



Just to make clear, after looking at SDL_gesture.c, what it does is:

- "$1" gesture. You draw something with one finger and (or mouse if synthesizing). And it can detect the gesture comparing to what it knows, giving the best diff result. User has to record his set of gesture before. 
 (http://depts.washington.edu/acelab/proj/dollar/index.html)

- It also generates what it calls "multi-gesture" event, when several fingers are pressed. It sends SDL events with the basic information:
 centroid, distance, and the rotation vs previous position. which sounds equivalent pinch in/out and to gesture rotation. (like in https://wayland.freedesktop.org/libinput/doc/latest/gestures.html )

- It seems there are also other algorithm, beside $1 which is 1 finger, $n, $p ... So there is a room for improvement ...

( I am not really impacted, because I have already disabled use them, and/or I re-implemented them. ) 
I don't think people need more than pinch, scroll, rotation, but may be wrong.



If needed, I can continue the PR https://github.com/libsdl-org/SDL/pull/6790 to move SDL_gesture.c appart.
either:
1/remove it totally (that could be a new pr infact)
or 2/ moving it appart, but kept in SDL, like:
- add event to be aware when touch device are added.
- expose SDL_GestureInit / Quit  in public header so that SDL user would have to manually call them if they want gesture.
or 3/ moving it appart like 2/:
- add event to be aware when touch device are added.
- remove SDL_gesture.c from SDL/src/ and move it to test.
- so that testgesture call SDL_GestureInit / Quit to does it's gesture

**Comment by icculus** [comment]
It might not be an issue, but I'd be concerned that a generic pinch/zoom/scroll implementation is going to make iOS and Android users go "this feels wrong" because everything else depends on the OS to decide what this feels like.

(Exposing those gestures from the OS through SDL? That might be fine, but I haven't thought about it in detail at all.)

But for now, am I clear to pull the existing gesture stuff out of here and into sdl2-compat (and/or an external library)? I don't want to do it yet if there's still some disagreement about the API's direction here.

**Comment by slouken** [comment]
Yes, you're clear. I think whatever we do for OS gesture events will be different from the existing API.

**Comment by 1bsyl** [comment]
to remove the gesture from SDL3, it's straight forward.

but to re-added into sdl2-compat,  SDL3 could export event when touch device are added.
need here: 
https://github.com/libsdl-org/SDL/blob/main/src/events/SDL_touch.c#L193
https://github.com/libsdl-org/SDL/blob/main/src/events/SDL_touch.c#L488
it says it could be added on the fly, but not sure if this is the best.

**Comment by icculus** [comment]
Okay, it's gone from SDL3 in https://github.com/libsdl-org/SDL/commit/a76053352c7d04a711260c3417f664c6da7a13f3, and added to sdl2-compat in https://github.com/libsdl-org/sdl2-compat/commit/d14dac9da48ea59f30dc272029c9d3a0972e6683.

>  it says it could be added on the fly, but not sure if this is the best.

Right now it adds touch devices as it sees events from them (and syncs up with SDL3 completely when the app calls SDL_RecordGesture), which is good enough for now.

But we can totally add an event for this to SDL3 anyhow!

**Comment by cblc** [comment]
I know this is closed, but just my 2 cents saying that I really need support for the OS-standard pan/zoom/rotate trackpad gestures on Mac and iOS. And I thought that SDL2 already had support for them, but looking at the source again, I believe my eyes saw what they wanted to see when I checked this last year.

Do you know of any PR/fork implementing this? I dislike having to use a custom modification of SDL, but I really need this feature...

**Comment by icculus** [comment]
I think an interface to the OS-provided pinch/zoom gestures makes sense. I'll open up a separate issue for that and we can discuss further.

**Comment by cblc** [comment]
Do you think there would be a chance for accepting a PR for this in SDL2, or would it have to be posponed for SDL3? I wrote my own (Mac-only) implementation of this for wxWidgets a long time ago (before they supported it), and I think I could contribute a PR if there's unanimity on how the API should be. I have no idea on how this works on Windows or Linux, though.
